{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1f88f55c-3150-462a-b73d-7cccfdfd6a49",
   "metadata": {},
   "source": [
    "### Circos plot m6A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fdda2ad2-b168-467d-8fe4-8e5bdd6bfa82",
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert predictions to bedgraph\n",
    "\n",
    "def write_filtered_bedgraph(input_file, output_file, min_depth, min_mod_rate):\n",
    "    with open(output_file, \"w\") as out:\n",
    "        with open(input_file) as f:\n",
    "            for line in f:\n",
    "                line = line.rstrip()\n",
    "                fields = line.split(\"\\t\")\n",
    "                chr = fields[3]\n",
    "                start = fields[4]\n",
    "                end = fields[4]\n",
    "                depth = int(fields[12])\n",
    "                mod_rate = int(fields[11]) / int(fields[12])\n",
    "                if depth > min_depth and mod_rate > min_mod_rate:\n",
    "                    out.write(\"%s\\t%s\\t%s\\t%s\\n\" % (chr, start, end, mod_rate))\n",
    "\n",
    "# Write filtered bedgraph for WT_m6A\n",
    "norm_factor=2\n",
    "input_file_wt = \"/home/wuyou/Projects/paper/rice/predict/WT_m6A.predict.genome_loc.merge.tsv\"\n",
    "output_file_wt = \"/home/wuyou/Projects/paper/rice/predict/WT_m6A.predict.bedgraph\"\n",
    "min_depth_wt = 5 * norm_factor\n",
    "min_mod_rate_wt = 0.3\n",
    "write_filtered_bedgraph(input_file_wt, output_file_wt, min_depth_wt, min_mod_rate_wt)\n",
    "\n",
    "# Write filtered bedgraph for NaCl_m6A\n",
    "norm_factor=1\n",
    "input_file_nacl = \"/home/wuyou/Projects/paper/rice/predict/NaCl_m6A.predict.genome_loc.merge.tsv\"\n",
    "output_file_nacl = \"/home/wuyou/Projects/paper/rice/predict/NaCl_m6A.predict.bedgraph\"\n",
    "min_depth_nacl = 5 * norm_factor\n",
    "min_mod_rate_nacl = 0.3\n",
    "write_filtered_bedgraph(input_file_nacl, output_file_nacl, min_depth_nacl, min_mod_rate_nacl)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8b64880-5902-4405-9f6d-ae0ec82aa9a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gene bin count\n",
    "import os\n",
    "\n",
    "max_length_dict={}\n",
    "with open(\"/home/wuyou//Projects/Osativa_nanopore/ref/Osativa_323_v7.0.chrom.sizes\") as f:\n",
    "    for line in f:\n",
    "        line=line.rstrip()\n",
    "        if \"ChrUn\" in line:\n",
    "            continue\n",
    "        if \"ChrSy\" in line:\n",
    "            continue\n",
    "        chr=line.split(\"\\t\")[0]\n",
    "        max_length=int(line.split(\"\\t\")[1])\n",
    "        max_length_dict[chr]=max_length\n",
    "\n",
    "#chromosome bin\n",
    "with open(\"/home/wuyou/Projects/paper/rice/circos/gene_bin.bed\",\"w\") as f:\n",
    "    bin_size=1e6\n",
    "    for chr in max_length_dict:\n",
    "        for i in range(int(max_length_dict[chr]/bin_size+1)):\n",
    "            f.writelines(\"%s\\t%s\\t%s\\n\" %(chr,int(i*bin_size),int((i+1)*bin_size)))\n",
    "            \n",
    "#gtf to bed\n",
    "out=open(\"/home/wuyou/Projects/paper/rice/circos/ref_gene.bed\",\"w\")\n",
    "with open(\"/home/wuyou//Projects/Osativa_nanopore/ref/Osativa_323_v7.0.gene_exons.gff3\") as f:\n",
    "    for line in f:\n",
    "        if \"ChrUn\" in line:\n",
    "            continue\n",
    "        if \"gene\" in line:\n",
    "            chr=line.split(\"\\t\")[0]\n",
    "            start=line.split(\"\\t\")[3]\n",
    "            end=line.split(\"\\t\")[4]\n",
    "            out.writelines(\"%s\\t%s\\t%s\\n\" %(chr,start,end))\n",
    "            \n",
    "cmd=\"bedtools intersect -a /home/wuyou/Projects/paper/rice/circos/gene_bin.bed -b  /home/wuyou/Projects/paper/rice/circos/ref_gene.bed -wa -c >/home/wuyou/Projects/paper/rice/circos/gene_bin_count.bed\"            \n",
    "print(cmd)\n",
    "os.system(cmd)        \n",
    "        \n",
    "\n",
    "#Prediction to bed\n",
    "#with open(\"/home/wuyou/Projects/paper/rice/predict/WT_m6A.predict.genome_loc.merge_5_0_3.bedgraph\")\n",
    "cmd=\"bedtools intersect -a /home/wuyou/Projects/paper/rice/circos/gene_bin.bed -b /home/wuyou/Projects/paper/rice/predict/WT_m6A.predict.genome_loc.merge.bedgraph -wa -c >/home/wuyou/Projects/paper/rice/circos/WT_m6A_bin_count.bed\"\n",
    "os.system(cmd)\n",
    "cmd=\"bedtools intersect -a /home/wuyou/Projects/paper/rice/circos/gene_bin.bed -b /home/wuyou/Projects/paper/rice/predict/NaCl_m6A.predict.genome_loc.merge.bedgraph -wa -c >/home/wuyou/Projects/paper/rice/circos/NaCl_m6A_bin_count.bed\"\n",
    "os.system(cmd)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09a78355-8b55-4e7f-91cc-22084742fa26",
   "metadata": {},
   "outputs": [],
   "source": [
    "#centromere\n",
    "with open(\"rice/circos/centromere.bed\",\"w\") as f:\n",
    "    f.writelines(\"Chr1\\t17308000\\t18960000\\t1\\n\")\n",
    "    f.writelines(\"Chr2\\t13893000\\t14171000\\t1\\n\")\n",
    "    f.writelines(\"Chr3\\t20631000\\t21560000\\t1\\n\")\n",
    "    f.writelines(\"Chr4\\t9991000\\t11794000\\t1\\n\")\n",
    "    f.writelines(\"Chr5\\t11850000\\t12898000\\t1\\n\")\n",
    "    f.writelines(\"Chr6\\t14757000\\t15952000\\t1\\n\")\n",
    "    f.writelines(\"Chr7\\t12783000\\t13473000\\t1\\n\")\n",
    "    f.writelines(\"Chr8\\t12555000\\t14323000\\t1\\n\")\n",
    "    f.writelines(\"Chr9\\t2505000\\t3107000\\t1\\n\")\n",
    "    f.writelines(\"Chr10\\t9628000\\t10201000\\t1\\n\")\n",
    "    f.writelines(\"Chr11\\t12488000\\t13992000\\t1\\n\")\n",
    "    f.writelines(\"Chr12\\t10607000\\t11634000\\t1\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "632df9f8-9b85-49fc-b47a-89e874612ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Rcode\n",
    "\n",
    "library(circlize)\n",
    "library(\"extrafont\")\n",
    "#font_import()\n",
    "\n",
    "black = '#222222'\n",
    "gray = '#666666'\n",
    "red = '#FF3333'\n",
    "green = '#66CC00'\n",
    "blue = '#3333FF'\n",
    "purple = '#9933FF'\n",
    "orange = '#FF8000'\n",
    "yellow = '#FFFF33'\n",
    "salmen = \"#FA8072\"\n",
    "limegreen = '#32CD32'\n",
    "deepskyblue = '#00AFEE'\n",
    "c1=\"#F8766D\"\n",
    "c2=\"#00BA38\"\n",
    "c3=\"#619CFF\"\n",
    "\n",
    "\n",
    "#pdf(file='circos.pdf',width=600,height=600)\n",
    "pdf(\"figure/figure_6a_circos_rice_m6A.pdf\",width=4/2.54, height=4/2.54,pointsize=8)\n",
    "ref=read.table(\"/home/wuyou/Projects/Osativa_nanopore/ref/Osativa_323_v7.0.chrom.sizes\",header=FALSE) ## length of chromosomes\n",
    "ref=head(ref,12)\n",
    "\n",
    "\n",
    "colnames(ref)=c(\"Genome\",\"Length\")\n",
    "\n",
    "ref$Genome=factor(c(\"Chr1\",\"Chr2\",\"Chr3\",\"Chr4\",\"Chr5\",\"Chr6\",\"Chr7\",\"Chr8\",\"Chr9\",\"Chr10\",\"Chr11\",\"Chr12\"),levels=c(\"Chr1\",\"Chr2\",\"Chr3\",\"Chr4\",\"Chr5\",\"Chr6\",\"Chr7\",\"Chr8\",\"Chr9\",\"Chr10\",\"Chr11\",\"Chr12\"))\n",
    "\n",
    "## create an initial/basic circle\n",
    "## set genome breaks, facing angle, fonts and so on\n",
    "circos.clear()\n",
    "col_text <- \"grey20\"\n",
    "circos.par(\"track.height\"=0.8,gap.after=c(1,1,1,1,1,1,1,1,1,1,1,10),start.degree =85,clock.wise = T,\n",
    "           cell.padding=c(0,0,0,0))\n",
    "circos.initialize(factors=ref$Genome,\t\t\t\t\t\t\t\t\t\t### see \n",
    "                  xlim=matrix(c(rep(0,12),ref$Length),ncol=2))\n",
    "\n",
    "\n",
    "#chr labels\n",
    "circos.track(ylim=c(0,1),panel.fun=function(x,y) {\n",
    "  Genome=CELL_META$sector.index\n",
    "  xlim=CELL_META$xlim\n",
    "  ylim=CELL_META$ylim\n",
    "  circos.text(mean(xlim),mean(ylim),Genome,cex=0.25,font=2,col=col_text,\n",
    "              facing=\"outside\",niceFacing=TRUE)\n",
    "},bg.col=\"white\",bg.border=F,track.height=0.08)\n",
    "\n",
    "#chr full length\n",
    "circos.track(ylim=c(0,1),panel.fun=function(x,y) {\n",
    "  Genome=CELL_META$sector.index\n",
    "  xlim=CELL_META$xlim\n",
    "  ylim=CELL_META$ylim\n",
    "},bg.col=c1,bg.border=T,track.height=0.06,bg.lwd=0.1)\n",
    "\n",
    "\n",
    "if (FALSE){\n",
    "#size label\n",
    "brk <- c(0,200,400,600,800,1000)*10^5\n",
    "circos.track(track.index = get.current.track.index(), panel.fun = function(x, y) {\n",
    "  circos.axis(h=\"top\",major.at=brk,labels=round(brk/10^5,1),labels.cex=0.1,\n",
    "              col=col_text,labels.col=col_text,lwd=0.1,labels.facing=\"outside\")\n",
    "},bg.border=F,track.height=0.06)\n",
    "\n",
    "}\n",
    "\n",
    "#centrometer\n",
    "centrometer<-read.table(\"rice/circos/centromere.bed\",header=FALSE) ## length of chromosomes\n",
    "colnames(centrometer)=c(\"chr\",\"start\",\"end\",\"value\")\n",
    "col_fun = colorRamp2(c(0, 1), c(\"white\", \"black\"))\n",
    "circos.genomicTrackPlotRegion(track.index = get.current.track.index(),\n",
    "  centrometer, track.height = 0.08, stack = TRUE, bg.border = NA,\n",
    "  panel.fun = function(region, value, ...) {\n",
    "    circos.genomicRect(region, value, col=\"black\", border = NA, ...)\n",
    "  } )\n",
    "\n",
    "\n",
    "#gene density\n",
    "\n",
    "gene =read.table(\"/home/wuyou/Projects/paper/rice/circos/gene_bin_count.bed\",header=FALSE) ## length of chromosomes\n",
    "colnames(gene)=c(\"chr\",\"start\",\"end\",\"value1\")\n",
    "col_fun = colorRamp2(c(0, 150), c(\"white\", \"#FF8C00\"))\n",
    "\n",
    "circos.genomicTrackPlotRegion(\n",
    "  gene, track.height = 0.04, stack = TRUE, bg.border = NA,\n",
    "  panel.fun = function(region, value, ...) {\n",
    "    circos.genomicRect(region, value, col=col_fun(value[[1]]), border = NA)\n",
    "  })\n",
    "\n",
    "\n",
    "#WT_m6A\n",
    "WT_m6A =read.table(\"/home/wuyou/Projects/paper/rice/circos/WT_m6A_bin_count.bed\",header=FALSE) ## length of chromosomes\n",
    "colnames(WT_m6A)=c(\"chr\",\"start\",\"end\",\"value1\")\n",
    "col_fun = colorRamp2(c(0, 500), c(\"white\", \"#0000FF\"))\n",
    "WT_m6A$value1=WT_m6A$value1/max(WT_m6A$value1)\n",
    "circos.genomicTrackPlotRegion(\n",
    "  WT_m6A, ylim=c(0, 1), track.height = 0.12,bg.border = NA,panel.fun = function(region, value, ...) {\n",
    "#value/142.381*200\n",
    "    circos.genomicRect(region, value, ytop.column = 1,ybottom = 0,col='#008080', border = NA, ...)\n",
    "})\n",
    "\n",
    "#NaCl_m6A\n",
    "NaCl_m6A =read.table(\"/home/wuyou/Projects/paper/rice/circos/NaCl_m6A_bin_count.bed\",header=FALSE) ## length of chromosomes\n",
    "colnames(NaCl_m6A)=c(\"chr\",\"start\",\"end\",\"value1\")\n",
    "col_fun = colorRamp2(c(0, 500), c(\"white\", \"#0000FF\"))\n",
    "NaCl_m6A$value1=NaCl_m6A$value1/max(NaCl_m6A$value1)\n",
    "circos.genomicTrackPlotRegion(\n",
    "  NaCl_m6A, ylim=c(0, 1), track.height = 0.12,bg.border = NA,panel.fun = function(region, value, ...) {\n",
    "#value/142.381*200\n",
    "    circos.genomicRect(region, value, ytop.column = 1,ybottom = 0,col='#008080', border = NA, ...)\n",
    "})\n",
    "\n",
    "dev.off()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edb4819b-bdd7-4b1d-b084-0a712fa7ffa4",
   "metadata": {},
   "source": [
    "### Circos plot m5C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6e4f560-5fa1-4b63-89a9-031c00b274fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#prediction to bed graph\n",
    "\n",
    "def write_filtered_bedgraph(input_file, output_file, min_depth, min_mod_rate):\n",
    "    with open(output_file, \"w\") as out:\n",
    "        with open(input_file) as f:\n",
    "            for line in f:\n",
    "                line = line.rstrip()\n",
    "                fields = line.split(\"\\t\")\n",
    "                chr = fields[3]\n",
    "                start = fields[4]\n",
    "                end = fields[4]\n",
    "                depth = int(fields[12])\n",
    "                mod_rate = int(fields[11]) / int(fields[12])\n",
    "                if depth > min_depth and mod_rate > min_mod_rate:\n",
    "                    out.write(\"%s\\t%s\\t%s\\t%s\\n\" % (chr, start, end, mod_rate))\n",
    "\n",
    "# Write filtered bedgraph for WT_C\n",
    "norm_factor = 2\n",
    "input_file_wt_c = \"/home/wuyou/Projects/paper/rice/predict/WT_C.predict.genome_loc.merge.tsv\"\n",
    "output_file_wt_c = \"/home/wuyou/Projects/paper/rice/predict/WT_C.predict.bedgraph\"\n",
    "min_depth_wt_c = 5 * norm_factor\n",
    "min_mod_rate_wt_c = 0.3\n",
    "write_filtered_bedgraph(input_file_wt_c, output_file_wt_c, min_depth_wt_c, min_mod_rate_wt_c)\n",
    "\n",
    "# Write filtered bedgraph for NaCl_C\n",
    "norm_factor = 1\n",
    "input_file_nacl_c = \"/home/wuyou/Projects/paper/rice/predict/NaCl_C.predict.genome_loc.merge.tsv\"\n",
    "output_file_nacl_c = \"/home/wuyou/Projects/paper/rice/predict/NaCl_C.predict.bedgraph\"\n",
    "min_depth_nacl_c = 5 * norm_factor\n",
    "min_mod_rate_nacl_c = 0.3\n",
    "write_filtered_bedgraph(input_file_nacl_c, output_file_nacl_c, min_depth_nacl_c, min_mod_rate_nacl_c)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d65e2f7-372f-432f-859d-4061a6f4674f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gene bin count\n",
    "import os\n",
    "\n",
    "max_length_dict={}\n",
    "with open(\"/home/wuyou//Projects/Osativa_nanopore/ref/Osativa_323_v7.0.chrom.sizes\") as f:\n",
    "    for line in f:\n",
    "        line=line.rstrip()\n",
    "        if \"ChrUn\" in line:\n",
    "            continue\n",
    "        if \"ChrSy\" in line:\n",
    "            continue\n",
    "        chr=line.split(\"\\t\")[0]\n",
    "        max_length=int(line.split(\"\\t\")[1])\n",
    "        max_length_dict[chr]=max_length\n",
    "\n",
    "#chromosome bin\n",
    "\n",
    "with open(\"/home/wuyou/Projects/paper/rice/circos/gene_bin.bed\",\"w\") as f:\n",
    "    bin_size=1e6\n",
    "    for chr in max_length_dict:\n",
    "        for i in range(int(max_length_dict[chr]/bin_size+1)):\n",
    "            f.writelines(\"%s\\t%s\\t%s\\n\" %(chr,int(i*bin_size),int((i+1)*bin_size)))\n",
    "            \n",
    "#gtf to bed\n",
    "\n",
    "out=open(\"/home/wuyou/Projects/paper/rice/circos/ref_gene.bed\",\"w\")\n",
    "with open(\"/home/wuyou//Projects/Osativa_nanopore/ref/Osativa_323_v7.0.gene_exons.gff3\") as f:\n",
    "    for line in f:\n",
    "        if \"ChrUn\" in line:\n",
    "            continue\n",
    "        if \"ChrSy\" in line:\n",
    "            continue\n",
    "        if \"gene\" in line:\n",
    "            chr=line.split(\"\\t\")[0]\n",
    "            start=line.split(\"\\t\")[3]\n",
    "            end=line.split(\"\\t\")[4]\n",
    "            out.writelines(\"%s\\t%s\\t%s\\n\" %(chr,start,end))\n",
    "        \n",
    "cmd=\"bedtools intersect -a /home/wuyou/Projects/paper/rice/circos/gene_bin.bed -b  /home/wuyou/Projects/paper/rice/circos/ref_gene.bed -wa -c >/home/wuyou/Projects/paper/rice/circos/gene_bin_count.bed\"            \n",
    "print(cmd)\n",
    "#os.system(cmd)        \n",
    "        \n",
    "\n",
    "#Prediction to bed\n",
    "#with open(\"/home/wuyou/Projects/paper/rice/predict/WT_m6A.predict.genome_loc.merge_5_0_3.bedgraph\")\n",
    "cmd=\"bedtools intersect -a /home/wuyou/Projects/paper/rice/circos/gene_bin.bed -b /home/wuyou/Projects/paper/rice/predict/WT_C.predict.bedgraph -wa -c >/home/wuyou/Projects/paper/rice/circos/WT_m5C_bin_count.bed\"\n",
    "os.system(cmd)\n",
    "cmd=\"bedtools intersect -a /home/wuyou/Projects/paper/rice/circos/gene_bin.bed -b /home/wuyou/Projects/paper/rice/predict/NaCl_C.predict.bedgraph -wa -c >/home/wuyou/Projects/paper/rice/circos/NaCl_m5C_bin_count.bed\"\n",
    "os.system(cmd)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a8fa448-cd65-46c2-bcea-8a4abd181eff",
   "metadata": {},
   "outputs": [],
   "source": [
    "#centromere\n",
    "with open(\"rice/circos/centromere.bed\",\"w\") as f:\n",
    "    f.writelines(\"Chr1\\t17308000\\t18960000\\t1\\n\")\n",
    "    f.writelines(\"Chr2\\t13893000\\t14171000\\t1\\n\")\n",
    "    f.writelines(\"Chr3\\t20631000\\t21560000\\t1\\n\")\n",
    "    f.writelines(\"Chr4\\t9991000\\t11794000\\t1\\n\")\n",
    "    f.writelines(\"Chr5\\t11850000\\t12898000\\t1\\n\")\n",
    "    f.writelines(\"Chr6\\t14757000\\t15952000\\t1\\n\")\n",
    "    f.writelines(\"Chr7\\t12783000\\t13473000\\t1\\n\")\n",
    "    f.writelines(\"Chr8\\t12555000\\t14323000\\t1\\n\")\n",
    "    f.writelines(\"Chr9\\t2505000\\t3107000\\t1\\n\")\n",
    "    f.writelines(\"Chr10\\t9628000\\t10201000\\t1\\n\")\n",
    "    f.writelines(\"Chr11\\t12488000\\t13992000\\t1\\n\")\n",
    "    f.writelines(\"Chr12\\t10607000\\t11634000\\t1\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8880d3cf-4136-45ea-b613-0762a6f2143a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#R code\n",
    "\n",
    "library(circlize)\n",
    "\n",
    "library(\"extrafont\")\n",
    "\n",
    "\n",
    "black = '#222222'\n",
    "gray = '#666666'\n",
    "red = '#FF3333'\n",
    "green = '#66CC00'\n",
    "blue = '#3333FF'\n",
    "purple = '#9933FF'\n",
    "orange = '#FF8000'\n",
    "yellow = '#FFFF33'\n",
    "salmen = \"#FA8072\"\n",
    "limegreen = '#32CD32'\n",
    "deepskyblue = '#00AFEE'\n",
    "c1=\"#F8766D\"\n",
    "c2=\"#00BA38\"\n",
    "c3=\"#619CFF\"\n",
    "\n",
    "\n",
    "#pdf(file='circos.pdf',width=600,height=600)\n",
    "pdf(\"figure/figure_6a_circos_rice_m5C.pdf\",width=4/2.54, height=4/2.54,pointsize=8)\n",
    "ref=read.table(\"/home/wuyou/Projects/Osativa_nanopore/ref/Osativa_323_v7.0.chrom.sizes\",header=FALSE) ## length of chromosomes\n",
    "ref=head(ref,12)\n",
    "\n",
    "\n",
    "colnames(ref)=c(\"Genome\",\"Length\")\n",
    "\n",
    "ref$Genome=factor(c(\"Chr1\",\"Chr2\",\"Chr3\",\"Chr4\",\"Chr5\",\"Chr6\",\"Chr7\",\"Chr8\",\"Chr9\",\"Chr10\",\"Chr11\",\"Chr12\"),levels=c(\"Chr1\",\"Chr2\",\"Chr3\",\"Chr4\",\"Chr5\",\"Chr6\",\"Chr7\",\"Chr8\",\"Chr9\",\"Chr10\",\"Chr11\",\"Chr12\"))\n",
    "\n",
    "## create an initial/basic circle\n",
    "## set genome breaks, facing angle, fonts and so on\n",
    "circos.clear()\n",
    "col_text <- \"grey20\"\n",
    "circos.par(\"track.height\"=0.8,gap.after=c(1,1,1,1,1,1,1,1,1,1,1,10),start.degree =85,clock.wise = T,\n",
    "           cell.padding=c(0,0,0,0))\n",
    "circos.initialize(factors=ref$Genome,\t\t\t\t\t\t\t\t\t\t### see \n",
    "                  xlim=matrix(c(rep(0,12),ref$Length),ncol=2))\n",
    "\n",
    "\n",
    "#chr labels\n",
    "circos.track(ylim=c(0,1),panel.fun=function(x,y) {\n",
    "  Genome=CELL_META$sector.index\n",
    "  xlim=CELL_META$xlim\n",
    "  ylim=CELL_META$ylim\n",
    "  circos.text(mean(xlim),mean(ylim),Genome,cex=0.25,font=2,col=col_text,\n",
    "              facing=\"outside\",niceFacing=TRUE)\n",
    "},bg.col=\"white\",bg.border=F,track.height=0.08)\n",
    "\n",
    "#chr full length\n",
    "circos.track(ylim=c(0,1),panel.fun=function(x,y) {\n",
    "  Genome=CELL_META$sector.index\n",
    "  xlim=CELL_META$xlim\n",
    "  ylim=CELL_META$ylim\n",
    "},bg.col=c1,bg.border=T,track.height=0.06,bg.lwd=0.1)\n",
    "\n",
    "\n",
    "if (FALSE){\n",
    "#size label\n",
    "brk <- c(0,200,400,600,800,1000)*10^5\n",
    "circos.track(track.index = get.current.track.index(), panel.fun = function(x, y) {\n",
    "  circos.axis(h=\"top\",major.at=brk,labels=round(brk/10^5,1),labels.cex=0.1,\n",
    "              col=col_text,labels.col=col_text,lwd=0.1,labels.facing=\"outside\")\n",
    "},bg.border=F,track.height=0.06)\n",
    "\n",
    "}\n",
    "\n",
    "#centrometer\n",
    "centrometer<-read.table(\"rice/circos/centromere.bed\",header=FALSE) ## length of chromosomes\n",
    "colnames(centrometer)=c(\"chr\",\"start\",\"end\",\"value\")\n",
    "col_fun = colorRamp2(c(0, 1), c(\"white\", \"black\"))\n",
    "circos.genomicTrackPlotRegion(track.index = get.current.track.index(),\n",
    "  centrometer, track.height = 0.08, stack = TRUE, bg.border = NA,\n",
    "  panel.fun = function(region, value, ...) {\n",
    "    circos.genomicRect(region, value, col=\"black\", border = NA, ...)\n",
    "  } )\n",
    "\n",
    "\n",
    "\n",
    "#gene density\n",
    "\n",
    "gene =read.table(\"/home/wuyou/Projects/paper/rice/circos/gene_bin_count.bed\",header=FALSE) ## length of chromosomes\n",
    "colnames(gene)=c(\"chr\",\"start\",\"end\",\"value1\")\n",
    "col_fun = colorRamp2(c(0, 150), c(\"white\", \"#FF8C00\"))\n",
    "\n",
    "circos.genomicTrackPlotRegion(\n",
    "  gene, track.height = 0.04, stack = TRUE, bg.border = NA,\n",
    "  panel.fun = function(region, value, ...) {\n",
    "    circos.genomicRect(region, value, col=col_fun(value[[1]]), border = NA)\n",
    "  })\n",
    "\n",
    "\n",
    "#WT_m6A\n",
    "WT_m6A =read.table(\"/home/wuyou/Projects/paper/rice/circos/WT_m5C_bin_count.bed\",header=FALSE) ## length of chromosomes\n",
    "colnames(WT_m6A)=c(\"chr\",\"start\",\"end\",\"value1\")\n",
    "col_fun = colorRamp2(c(0, 500), c(\"white\", \"#0000FF\"))\n",
    "WT_m6A$value1=WT_m6A$value1/max(WT_m6A$value1)\n",
    "circos.genomicTrackPlotRegion(\n",
    "  WT_m6A, ylim=c(0, 1), track.height = 0.12,bg.border = NA,panel.fun = function(region, value, ...) {\n",
    "#value/142.381*200\n",
    "    circos.genomicRect(region, value, ytop.column = 1,ybottom = 0,col='#008080', border = NA, ...)\n",
    "})\n",
    "\n",
    "#NaCl_m6A\n",
    "NaCl_m6A =read.table(\"/home/wuyou/Projects/paper/rice/circos/NaCl_m5C_bin_count.bed\",header=FALSE) ## length of chromosomes\n",
    "colnames(NaCl_m6A)=c(\"chr\",\"start\",\"end\",\"value1\")\n",
    "col_fun = colorRamp2(c(0, 500), c(\"white\", \"#0000FF\"))\n",
    "NaCl_m6A$value1=NaCl_m6A$value1/max(NaCl_m6A$value1)\n",
    "circos.genomicTrackPlotRegion(\n",
    "  NaCl_m6A, ylim=c(0, 1), track.height = 0.12,bg.border = NA,panel.fun = function(region, value, ...) {\n",
    "#value/142.381*200\n",
    "    circos.genomicRect(region, value, ytop.column = 1,ybottom = 0,col='#008080', border = NA, ...)\n",
    "})\n",
    "\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0145532-f680-4ad9-acca-b11fbf67f25a",
   "metadata": {},
   "source": [
    "### Circos plot psU"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec3d5289-920f-4569-818e-a1c83acd0c21",
   "metadata": {},
   "outputs": [],
   "source": [
    "#predict to bed graph\n",
    "\n",
    "def write_filtered_bedgraph(input_file, output_file, min_depth, min_mod_rate):\n",
    "    with open(output_file, \"w\") as out:\n",
    "        with open(input_file) as f:\n",
    "            for line in f:\n",
    "                line = line.rstrip()\n",
    "                fields = line.split(\"\\t\")\n",
    "                chr = fields[3]\n",
    "                start = fields[4]\n",
    "                end = fields[4]\n",
    "                depth = int(fields[12])\n",
    "                mod_rate = int(fields[11]) / int(fields[12])\n",
    "                if depth > min_depth and mod_rate > min_mod_rate:\n",
    "                    out.write(\"%s\\t%s\\t%s\\t%s\\n\" % (chr, start, end, mod_rate))\n",
    "\n",
    "# Write filtered bedgraph for WT_U\n",
    "norm_factor = 2\n",
    "input_file_wt_u = \"/home/wuyou/Projects/paper/rice/results/WT_U.predict.genome_loc.merge.tsv\"\n",
    "output_file_wt_u = \"/home/wuyou/Projects/paper/rice/results/WT_U.predict.bedgraph\"\n",
    "min_depth_wt_u = 5 * norm_factor\n",
    "min_mod_rate_wt_u = 0.3\n",
    "write_filtered_bedgraph(input_file_wt_u, output_file_wt_u, min_depth_wt_u, min_mod_rate_wt_u)\n",
    "\n",
    "# Write filtered bedgraph for NaCl_U\n",
    "norm_factor = 1\n",
    "input_file_nacl_u = \"/home/wuyou/Projects/paper/rice/results/NaCl_U.predict.genome_loc.merge.tsv\"\n",
    "output_file_nacl_u = \"/home/wuyou/Projects/paper/rice/results/NaCl_U.predict.bedgraph\"\n",
    "min_depth_nacl_u = 5 * norm_factor\n",
    "min_mod_rate_nacl_u = 0.3\n",
    "write_filtered_bedgraph(input_file_nacl_u, output_file_nacl_u, min_depth_nacl_u, min_mod_rate_nacl_u)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bdc4bef-4f02-4725-b59d-866506dd856f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gene bin count\n",
    "import os\n",
    "\n",
    "max_length_dict={}\n",
    "with open(\"/home/wuyou//Projects/Osativa_nanopore/ref/Osativa_323_v7.0.chrom.sizes\") as f:\n",
    "    for line in f:\n",
    "        line=line.rstrip()\n",
    "        if \"ChrUn\" in line:\n",
    "            continue\n",
    "        if \"ChrSy\" in line:\n",
    "            continue\n",
    "        chr=line.split(\"\\t\")[0]\n",
    "        max_length=int(line.split(\"\\t\")[1])\n",
    "        max_length_dict[chr]=max_length\n",
    "\n",
    "#chromosome bin\n",
    "\n",
    "with open(\"/home/wuyou/Projects/paper/rice/circos/gene_bin.bed\",\"w\") as f:\n",
    "    bin_size=1e6\n",
    "    for chr in max_length_dict:\n",
    "        for i in range(int(max_length_dict[chr]/bin_size+1)):\n",
    "            f.writelines(\"%s\\t%s\\t%s\\n\" %(chr,int(i*bin_size),int((i+1)*bin_size)))\n",
    "            \n",
    "#gtf to bed\n",
    "\n",
    "out=open(\"/home/wuyou/Projects/paper/rice/circos/ref_gene.bed\",\"w\")\n",
    "with open(\"/home/wuyou//Projects/Osativa_nanopore/ref/Osativa_323_v7.0.gene_exons.gff3\") as f:\n",
    "    for line in f:\n",
    "        if \"ChrUn\" in line:\n",
    "            continue\n",
    "        if \"ChrSy\" in line:\n",
    "            continue\n",
    "        if \"gene\" in line:\n",
    "            chr=line.split(\"\\t\")[0]\n",
    "            start=line.split(\"\\t\")[3]\n",
    "            end=line.split(\"\\t\")[4]\n",
    "            out.writelines(\"%s\\t%s\\t%s\\n\" %(chr,start,end))\n",
    "        \n",
    "cmd=\"bedtools intersect -a /home/wuyou/Projects/paper/rice/circos/gene_bin.bed -b  /home/wuyou/Projects/paper/rice/circos/ref_gene.bed -wa -c >/home/wuyou/Projects/paper/rice/circos/gene_bin_count.bed\"            \n",
    "print(cmd)\n",
    "#os.system(cmd)        \n",
    "        \n",
    "\n",
    "#m6A prediction to bed\n",
    "cmd=\"bedtools intersect -a /home/wuyou/Projects/paper/rice/circos/gene_bin.bed -b /home/wuyou/Projects/paper/rice/results/WT_U.predict.bedgraph -wa -c >/home/wuyou/Projects/paper/rice/circos/WT_U_bin_count.bed\"\n",
    "os.system(cmd)\n",
    "cmd=\"bedtools intersect -a /home/wuyou/Projects/paper/rice/circos/gene_bin.bed -b /home/wuyou/Projects/paper/rice/results/NaCl_U.predict.bedgraph -wa -c >/home/wuyou/Projects/paper/rice/circos/NaCl_U_bin_count.bed\"\n",
    "os.system(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9ed8c52-e55b-4359-86aa-c8982852f9ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#centromere\n",
    "with open(\"rice/circos/centromere.bed\",\"w\") as f:\n",
    "    f.writelines(\"Chr1\\t17308000\\t18960000\\t1\\n\")\n",
    "    f.writelines(\"Chr2\\t13893000\\t14171000\\t1\\n\")\n",
    "    f.writelines(\"Chr3\\t20631000\\t21560000\\t1\\n\")\n",
    "    f.writelines(\"Chr4\\t9991000\\t11794000\\t1\\n\")\n",
    "    f.writelines(\"Chr5\\t11850000\\t12898000\\t1\\n\")\n",
    "    f.writelines(\"Chr6\\t14757000\\t15952000\\t1\\n\")\n",
    "    f.writelines(\"Chr7\\t12783000\\t13473000\\t1\\n\")\n",
    "    f.writelines(\"Chr8\\t12555000\\t14323000\\t1\\n\")\n",
    "    f.writelines(\"Chr9\\t2505000\\t3107000\\t1\\n\")\n",
    "    f.writelines(\"Chr10\\t9628000\\t10201000\\t1\\n\")\n",
    "    f.writelines(\"Chr11\\t12488000\\t13992000\\t1\\n\")\n",
    "    f.writelines(\"Chr12\\t10607000\\t11634000\\t1\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e4f136b-e7bf-4b35-bdae-f113b9e9bd89",
   "metadata": {},
   "outputs": [],
   "source": [
    "#R code\n",
    "\n",
    "library(circlize)\n",
    "\n",
    "library(\"extrafont\")\n",
    "#font_import()\n",
    "#fonts()\n",
    "#loadfonts()   #for pdf\n",
    "\n",
    "black = '#222222'\n",
    "gray = '#666666'\n",
    "red = '#FF3333'\n",
    "green = '#66CC00'\n",
    "blue = '#3333FF'\n",
    "purple = '#9933FF'\n",
    "orange = '#FF8000'\n",
    "yellow = '#FFFF33'\n",
    "salmen = \"#FA8072\"\n",
    "limegreen = '#32CD32'\n",
    "deepskyblue = '#00AFEE'\n",
    "c1=\"#F8766D\"\n",
    "c2=\"#00BA38\"\n",
    "c3=\"#619CFF\"\n",
    "\n",
    "\n",
    "#pdf(file='circos.pdf',width=600,height=600)\n",
    "pdf(\"figure/figure_6a_circos_rice_psU.pdf\",width=4/2.54, height=4/2.54,pointsize=8)\n",
    "ref=read.table(\"/home/wuyou/Projects/Osativa_nanopore/ref/Osativa_323_v7.0.chrom.sizes\",header=FALSE) ## length of chromosomes\n",
    "ref=head(ref,12)\n",
    "\n",
    "\n",
    "colnames(ref)=c(\"Genome\",\"Length\")\n",
    "\n",
    "ref$Genome=factor(c(\"Chr1\",\"Chr2\",\"Chr3\",\"Chr4\",\"Chr5\",\"Chr6\",\"Chr7\",\"Chr8\",\"Chr9\",\"Chr10\",\"Chr11\",\"Chr12\"),levels=c(\"Chr1\",\"Chr2\",\"Chr3\",\"Chr4\",\"Chr5\",\"Chr6\",\"Chr7\",\"Chr8\",\"Chr9\",\"Chr10\",\"Chr11\",\"Chr12\"))\n",
    "\n",
    "## create an initial/basic circle\n",
    "## set genome breaks, facing angle, fonts and so on\n",
    "circos.clear()\n",
    "col_text <- \"grey20\"\n",
    "circos.par(\"track.height\"=0.8,gap.after=c(1,1,1,1,1,1,1,1,1,1,1,10),start.degree =85,clock.wise = T,\n",
    "           cell.padding=c(0,0,0,0))\n",
    "circos.initialize(factors=ref$Genome,\t\t\t\t\t\t\t\t\t\t### see \n",
    "                  xlim=matrix(c(rep(0,12),ref$Length),ncol=2))\n",
    "\n",
    "\n",
    "#chr labels\n",
    "circos.track(ylim=c(0,1),panel.fun=function(x,y) {\n",
    "  Genome=CELL_META$sector.index\n",
    "  xlim=CELL_META$xlim\n",
    "  ylim=CELL_META$ylim\n",
    "  circos.text(mean(xlim),mean(ylim),Genome,cex=0.25,font=2,col=col_text,\n",
    "              facing=\"outside\",niceFacing=TRUE)\n",
    "},bg.col=\"white\",bg.border=F,track.height=0.08)\n",
    "\n",
    "#chr full length\n",
    "circos.track(ylim=c(0,1),panel.fun=function(x,y) {\n",
    "  Genome=CELL_META$sector.index\n",
    "  xlim=CELL_META$xlim\n",
    "  ylim=CELL_META$ylim\n",
    "},bg.col=c1,bg.border=T,track.height=0.06,bg.lwd=0.1)\n",
    "\n",
    "\n",
    "if (FALSE){\n",
    "#size label\n",
    "brk <- c(0,200,400,600,800,1000)*10^5\n",
    "circos.track(track.index = get.current.track.index(), panel.fun = function(x, y) {\n",
    "  circos.axis(h=\"top\",major.at=brk,labels=round(brk/10^5,1),labels.cex=0.1,\n",
    "              col=col_text,labels.col=col_text,lwd=0.1,labels.facing=\"outside\")\n",
    "},bg.border=F,track.height=0.06)\n",
    "\n",
    "}\n",
    "\n",
    "#centrometer\n",
    "centrometer<-read.table(\"rice/circos/centromere.bed\",header=FALSE) ## length of chromosomes\n",
    "colnames(centrometer)=c(\"chr\",\"start\",\"end\",\"value\")\n",
    "col_fun = colorRamp2(c(0, 1), c(\"white\", \"black\"))\n",
    "circos.genomicTrackPlotRegion(track.index = get.current.track.index(),\n",
    "  centrometer, track.height = 0.08, stack = TRUE, bg.border = NA,\n",
    "  panel.fun = function(region, value, ...) {\n",
    "    circos.genomicRect(region, value, col=\"black\", border = NA, ...)\n",
    "  } )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#gene density\n",
    "\n",
    "gene =read.table(\"/home/wuyou/Projects/paper/rice/circos/gene_bin_count.bed\",header=FALSE) ## length of chromosomes\n",
    "colnames(gene)=c(\"chr\",\"start\",\"end\",\"value1\")\n",
    "col_fun = colorRamp2(c(0, 150), c(\"white\", \"#FF8C00\"))\n",
    "\n",
    "circos.genomicTrackPlotRegion(\n",
    "  gene, track.height = 0.04, stack = TRUE, bg.border = NA,\n",
    "  panel.fun = function(region, value, ...) {\n",
    "    circos.genomicRect(region, value, col=col_fun(value[[1]]), border = NA)\n",
    "  })\n",
    "\n",
    "\n",
    "#WT_m6A\n",
    "WT_m6A =read.table(\"/home/wuyou/Projects/paper/rice/circos/WT_U_bin_count.bed\",header=FALSE) ## length of chromosomes\n",
    "colnames(WT_m6A)=c(\"chr\",\"start\",\"end\",\"value1\")\n",
    "col_fun = colorRamp2(c(0, 500), c(\"white\", \"#0000FF\"))\n",
    "WT_m6A$value1=WT_m6A$value1/max(WT_m6A$value1)\n",
    "circos.genomicTrackPlotRegion(\n",
    "  WT_m6A, ylim=c(0, 1), track.height = 0.12,bg.border = NA,panel.fun = function(region, value, ...) {\n",
    "#value/142.381*200\n",
    "    circos.genomicRect(region, value, ytop.column = 1,ybottom = 0,col='#008080', border = NA, ...)\n",
    "})\n",
    "\n",
    "#NaCl_m6A\n",
    "NaCl_m6A =read.table(\"/home/wuyou/Projects/paper/rice/circos/NaCl_U_bin_count.bed\",header=FALSE) ## length of chromosomes\n",
    "colnames(NaCl_m6A)=c(\"chr\",\"start\",\"end\",\"value1\")\n",
    "col_fun = colorRamp2(c(0, 500), c(\"white\", \"#0000FF\"))\n",
    "NaCl_m6A$value1=NaCl_m6A$value1/max(NaCl_m6A$value1)\n",
    "circos.genomicTrackPlotRegion(\n",
    "  NaCl_m6A, ylim=c(0, 1), track.height = 0.12,bg.border = NA,panel.fun = function(region, value, ...) {\n",
    "#value/142.381*200\n",
    "    circos.genomicRect(region, value, ytop.column = 1,ybottom = 0,col='#008080', border = NA, ...)\n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "dev.off()\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
